discord/bot.py

```diff
@@ -217,48 +217,94 @@ async def register_commands(self) -> None:
-                ]
-                # TODO: rewrite this, it seems inefficient
-                if matches:
-                    as_dict["id"] = matches[0]["id"]
-            commands.append(as_dict)
-
-        cmds = await self.http.bulk_upsert_global_commands(self.user.id, commands)
-
-        for i in cmds:
+        # 'Your app cannot have two global commands with the same name. Your app cannot have two guild commands within the same name on the same guild.'
+        # We can therefore safely use the name of the command in our global slash commands as a unique identifier
+        registered_commands_dict = {cmd["name"]:cmd for cmd in registered_commands}
+        global_pending_application_commands_dict = {}
+        
+        if len(registered_commands) > 0:
+            for command in [
+                cmd for cmd in self.pending_application_commands if cmd.guild_ids is None
+            ]:
+                as_dict = command.to_dict()
+                
+                global_pending_application_commands_dict[command.name] = as_dict
+                if command.name in registered_commands_dict:
+                    match = registered_commands_dict[command.name]
+                else:
+                    match = None
+                # TODO: There is probably a far more efficient way of doing this
+                # We want to check if the registered global command on Discord servers matches the given global commands
+                if match:
+                    as_dict["id"] = match["id"]
+
+                    keys_to_check = {"default_permission": True, "name": True, "description": True, "options": ["type", "name", "description"]}
+                    for key, more_keys in {
+                        key:more_keys
+                        for key, more_keys in keys_to_check.items()
+                        if key in as_dict.keys()
+                        if key in match.keys()
+                    }.items():
+                        if key == "options":
+                            for i, option_dict in enumerate(as_dict[key]):
+                                for key2_change in [
+                                    key2
+                                    for key2 in more_keys
+                                    if key2 in option_dict.keys() and key2 in match[key][i].keys()
+                                    if option_dict[key2] != match[key][i][key2]
+                                ]:
+                                    print(f"A change in the '{key2_change}' property of '{key}' in the '{command.name}' global command was noticed, will send to Discord to update")
+                                    needs_bulk = True
+                        else:
+                            if as_dict[key] != match[key]:
+                                print(f"A change in the '{key}' property of '{command.name}' global command was noticed, will send to Discord to update")
```

Comment:
We should probably send this to `logging.info` or else not log it at all. No need to print this.
```suggestion
                                logging.info(f"A change in the '{key}' property of '{command.name}' global command was noticed, will send to Discord to update")
```